File: ruby/lib/runapi/qwen_2/resources/edit_image.rb
# frozen_string_literal: true

module RunApi
  module Qwen2
    module Resources
      # Qwen2 edit-image resource. Edits images with natural-language prompts.
      class EditImage
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/qwen_2/edit_image"

        RESPONSE_CLASS = Types::EditImageResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedEditImageResponse

        def initialize(http)
          @http = http
        end

        # Edit an image and wait until complete.
        #
        # @param params [Hash] edit-image parameters
        # @return [RunApi::Qwen2::Types::CompletedEditImageResponse] completed edit-image result
        def run(**params)
          task = create(**params)
          poll_until_complete { get(task.id) }
        end

        # Create an edit-image task.
        #
        # @param params [Hash] edit-image parameters
        # @return [RunApi::Qwen2::Types::EditImageResponse] task creation result with id
        def create(**params)
          params = compact_params(params)
          validate_contract!(CONTRACT["edit-image"], params)
          request(:post, ENDPOINT, body: params)
        end

        # Get edit-image status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::Qwen2::Types::EditImageResponse] current edit-image status
        def get(id)
          request(:get, "#{ENDPOINT}/#{id}")
        end
      end
    end
  end
end

File: python/src/runapi/qwen_2/resources/edit_image.py
"""Qwen 2 edit-image resource."""

from __future__ import annotations

from typing import Any, Dict

from runapi.core import Resource, ValidationError

from ..types import (
    EDIT_IMAGE_ASPECT_RATIOS,
    EDIT_MODELS,
    OUTPUT_FORMATS,
    CompletedEditImageResponse,
    EditImageResponse,
)


class EditImage(Resource):
    """Edit images with natural-language prompts."""

    ENDPOINT = "/api/v1/qwen_2/edit_image"

    RESPONSE_CLASS = EditImageResponse
    COMPLETED_RESPONSE_CLASS = CompletedEditImageResponse

    def run(self, **params: Any) -> Any:
        """Create an edit-image task and poll until it completes.

        Args:
            **params: Edit-image parameters (model, prompt, source_image_url, ...).

        Returns:
            The completed (narrowed) response.
        """
        task = self.create(**params)
        return self._poll_until_complete(lambda: self.get(task.id))

    def create(self, **params: Any) -> Any:
        """Create an edit-image task and return immediately with an ``id``.

        Args:
            **params: Edit-image parameters (model, prompt, source_image_url, ...).

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted)

    def get(self, id: str) -> Any:
        """Fetch the current status of an edit-image task.

        Args:
            id: The task id.

        Returns:
            The current status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}")

    def _validate_params(self, params: Dict[str, Any]) -> None:
        if not params.get("model"):
            raise ValidationError("model is required")
        if not params.get("prompt"):
            raise ValidationError("prompt is required")
        if not params.get("source_image_url"):
            raise ValidationError("source_image_url is required")

        model = params.get("model")
        if model not in EDIT_MODELS:
            raise ValidationError(f"Invalid model: {model}. Must be one of: {', '.join(EDIT_MODELS)}")

        self._validate_optional(params, "aspect_ratio", EDIT_IMAGE_ASPECT_RATIOS)
        self._validate_optional(params, "output_format", OUTPUT_FORMATS)
